cli/src/main/java/com/liferay/blade/cli/command/WatchCommand.java

```diff
@@ -75,102 +98,307 @@ private void _addError(String prefix, String msg) {
+			watchPath,
+			new SimpleFileVisitor<Path>() {
 
-						bladeCLI.out(response);
+				@Override
+				public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					boolean shouldIgnorePath = ignorePathMatchers.stream(
+					).filter(
+						pathMatcher -> pathMatcher.matches(path)
+					).findFirst(
+					).isPresent();
+
+					if (shouldIgnorePath) {
+						return FileVisitResult.SKIP_SUBTREE;
 					}
-					catch (IOException ioe) {
+
+					try (Stream<Path> files = Files.list(path)) {
+						boolean projectPathFound = files.map(
+							p -> p.getFileName()
+						).filter(
+							p -> !ignorePathMatchers.stream(
+							).filter(
+								pathMatcher -> pathMatcher.matches(path.resolve(p))
+							).findFirst(
+							).isPresent()
+						).filter(
+							p -> projectPaths.stream(
+							).filter(
+								pp -> Objects.equals(pp, p.toString())
+							).findFirst(
+							).isPresent()
+						).findFirst(
+						).isPresent();
+
+						if (projectPathFound) {
+							foundProjectPaths.put(_getGradlePath(path, watchPath), path);
+
+							return FileVisitResult.SKIP_SUBTREE;
+						}
 					}
+
+					return FileVisitResult.CONTINUE;
 				}
-			}
-		}
-		catch (IOException ioe) {
-		}
+
+			});
+
+		return foundProjectPaths;
 	}
 
-	private void _watch(Path watchPath, Map<String, Set<File>> projectOutputFiles) throws InterruptedException {
+	private void _registerDirectory(WatchService watcher, Map<WatchKey, Path> keys, Path dir) throws IOException {
+		WatchKey watchKey = dir.register(
+			watcher,
+			new WatchEvent.Kind[] {
+				StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE,
+				StandardWatchEventKinds.ENTRY_MODIFY
+			},
+			SensitivityWatchEventModifier.HIGH);
+
+		keys.put(watchKey, dir);
+	}
+
+	private void _walkAndRegisterDirectories(
+			final WatchService watchService, final Map<WatchKey, Path> watchKeys, final Path basePath,
+			final List<PathMatcher> ignorePathMatchers)
+		throws IOException {
+
+		Files.walkFileTree(
+			basePath,
+			new SimpleFileVisitor<Path>() {
+
+				@Override
+				public FileVisitResult preVisitDirectory(Path path, BasicFileAttributes basicFileAttributes)
+					throws IOException {
+
+					for (PathMatcher pathMatcher : ignorePathMatchers) {
+						if (pathMatcher.matches(path)) {
+							return FileVisitResult.SKIP_SUBTREE;
+						}
+					}
+
+					_registerDirectory(watchService, watchKeys, path);
+
+					return FileVisitResult.CONTINUE;
+				}
+
+			});
+	}
+
+	private void _watch(
+			Path watchPath, Map<String, Path> projectPaths, List<String> fastPaths, List<String> ignorePaths,
+			boolean deploy)
+		throws InterruptedException {
+
 		Thread watchThread = new Thread() {
 
 			@Override
 			public void run() {
 				BladeCLI bladeCLI = getBladeCLI();
 
 				try {
+					FileSystem fileSystem = FileSystems.getDefault();
+
+					final WatchService watchService = fileSystem.newWatchService();
```

Comment:
`WatchService` should be wrapped in a try-with-resources block to ensure the `WatchService` object gets closed, as it implements `AutoCloseable`

Reply:
Thanks for tip.